module/x/peggy/abci.go

```diff
@@ -83,37 +176,60 @@ func slashing(ctx sdk.Context, k keeper.Keeper) {
 	} else if types.BridgeValidators(k.GetCurrentValset(ctx).Members).PowerDiff(latestValset.Members) > 0.05 {
 		k.SetValsetRequest(ctx)
 	}
+}
+
+func BatchSlashing(ctx sdk.Context, k keeper.Keeper, params types.Params, currentBondedSet []stakingtypes.Validator) {
 
 	// #2 condition
 	// We look through the full bonded set (not just the active set, include unbonding validators)
 	// and we slash users who haven't signed a batch confirmation that is >15hrs in blocks old
-	batches := k.GetOutgoingTxBatches(ctx)
-	for _, batch := range batches {
-		signedWithinWindow := uint64(ctx.BlockHeight()) > params.SignedBatchesWindow && uint64(ctx.BlockHeight())-params.SignedBatchesWindow > batch.Block
-		if signedWithinWindow {
-			confirms := k.GetBatchConfirmByNonceAndTokenContract(ctx, batch.BatchNonce, batch.TokenContract)
-			for _, val := range currentBondedSet {
-				found := false
-				for _, conf := range confirms {
-					// TODO: double check this logic
-					confVal, _ := sdk.AccAddressFromBech32(conf.Orchestrator)
-					if k.GetOrchestratorValidator(ctx, confVal).Equals(val.GetOperator()) {
-						found = true
-						break
-					}
-				}
-				if !found {
-					cons, _ := val.GetConsAddr()
-					k.StakingKeeper.Slash(ctx, cons, ctx.BlockHeight(), val.ConsensusPower(), params.SlashFractionBatch)
-					k.StakingKeeper.Jail(ctx, cons)
-				}
+	maxHeight := uint64(0)
+
+	// don't slash in the beginning before there aren't even SignedBatchesWindow blocks yet
+	if uint64(ctx.BlockHeight()) > params.SignedBatchesWindow {
+		maxHeight = uint64(ctx.BlockHeight()) - params.SignedBatchesWindow
+	}
+
+	unslashedBatches := k.GetUnSlashedBatches(ctx, maxHeight)
+	for _, batch := range unslashedBatches {
+
+		confirms := k.GetBatchConfirmByNonceAndTokenContract(ctx, batch.BatchNonce, batch.TokenContract)
+		for _, val := range currentBondedSet {
```

Comment:
It's actually questionable if the batch slashing really needs the current bonded set. Or any of this more complex logic. Unlike validator sets we don't really need every batch to succeed. If we can't get enough signatures because a validator left the set then we don't care, the batch will time out overnight anyways.

That being said all of this will work fine anyways and there are advantages to having the batch and validator set logic be the same. 

Just noting this for now, will merge this code anyways.